### ISSUE TITLE
Application signature check allocates large amount of RAM at install

### ISSUE THREAD

**Issue by David-5-1** [nextcloud/server#19463]
### Steps to reproduce
1. Configure PHP with `memory_limit = 256M` or anything lower than that.
2. Try to install Nextcloud 18 with recommended applications.
3. Wait for “Community Document Server” application installation to fail.

### Expected behaviour
All applications end up correctly installed.

### Actual behaviour
“Community Document Server” application install fails with a message in PHP logs (using PHP-FPM) like:
```
[Thu Feb 13 11:25:24.514471 2020] [proxy_fcgi:error] [pid 6126:tid 140382348179200] [client 217.70.181.1:60698] AH01071: Got error 'PHP message: PHP Fatal error:  Allowed memory size of 268435456 bytes exhausted (tried to allocate 315857416 bytes) in /var/www/nextcloud/lib/private/Installer.php on line 278'
```

The core of the issue as I understand it is the fact the Installer reads the whole application archive into memory through `file_get_contents($tempFile)` to check the signature as follow (re-indented for clarity):
```
$verified = (bool)openssl_verify(
    file_get_contents($tempFile),
    base64_decode($app['releases'][0]['signature']),
    $certificate,
    OPENSSL_ALGO_SHA512
);
```
https://github.com/nextcloud/server/blob/v18.0.1/lib/private/Installer.php#L278

I do not see any equivalent to `openssl_verify` in PHP that could take a file handle as an argument, hence the lack of an attached PR.

All the below information are IMO of no relevance for the issue but included anyway.

### Server configuration

**Operating system:** Debian Buster (or stretch)

**Web server:** Apache 2.4

**Database:** MariaDB 10.3

**PHP version:** 7.3

**Nextcloud version:** 18.0.1 (or 18.0.0)

**Updated from an older Nextcloud/ownCloud or fresh install:** fresh install

**Where did you install Nextcloud from:** https://download.nextcloud.com/server/releases/nextcloud-18.0.1.tar.bz2

**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
All the default ones for those installation worked…

Accessibility 1.4.0
Activity 2.11.0
Brute-force settings 1.5.0
Calendar 2.0.1
Collaborative tags 1.8.0
Comments 1.8.0
Contacts 3.1.8
Deleted files 1.8.0
Federation 1.8.0
File sharing 1.10.1
First run wizard 2.7.0
Log Reader 2.3.0
Mail 1.1.2
Monitoring 1.8.0
Nextcloud announcements 1.7.0
Notifications 2.6.0
ONLYOFFICE 4.1.4
Password policy 1.8.0
PDF viewer 1.7.0
Photos 1.0.0
Privacy 1.2.0
Recommendations 0.6.0
Right click 0.15.2
Share by mail 1.8.0
Support 1.1.0
Text 2.0.0
Theming 1.9.0
Update notification 1.8.0
Usage survey 1.6.0
Versions 1.11.0
Video player 1.7.0
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
<?php
$CONFIG = array (
  'instanceid' => 'XX',
  'passwordsalt' => 'XX',
  'secret' => 'XX',
  'trusted_domains' => 
  array (
    0 => 'XX',
  ),
  'datadirectory' => 'XX/data',
  'dbtype' => 'mysql',
  'version' => '18.0.1.3',
  'overwrite.cli.url' => 'XX',
  'dbname' => 'XX',
  'dbhost' => 'localhost',
  'dbport' => '',
  'dbtableprefix' => 'oc_',
  'mysql.utf8mb4' => true,
  'dbuser' => 'XX',
  'dbpassword' => 'XX',
  'installed' => true,
);

```
</details>

**Are you using external storage, if yes which one:** local

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** no

### Client configuration
**Browser:** Firefox 72

**Operating system:** Debian Buster

#### Web server error log
<details>
<summary>Web server error log</summary>

```
[Thu Feb 13 11:25:24.514471 2020] [proxy_fcgi:error] [pid 6126:tid 140382348179200] [client 217.70.181.1:60698] AH01071: Got error 'PHP message: PHP Fatal error:  Allowed memory size of 268435456 bytes exhausted (tried to allocate 315857416 bytes) in /var/www/nextcloud/lib/private/Installer.php on line 278'
```
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
None related to this issue.
```
</details>

**Comment by kesselb** [comment]
cc @rullzer that's indeed a issue.

**Comment by rullzer** [comment]
Right yes.
I think I have a solution for this. But it will require some hacking and work. I'll see if I can tackle this in the next few weeks.

**Comment by kesselb** [comment]
We could update the docs and state that for some apps 512mb is required. That's probably not that hacky ;) Do you think that the documentserver_community will work with 256mb per request?

**Comment by David-5-1** [comment]
Except for known oddities with file saving described in [document_server issue](https://github.com/nextcloud/documentserver_community/issues/12), I can get OnlyOffice to work correctly editing files with a 128 MB for PHP `memory_limit` as long as applications have been installed before lowering the value. (I tried both smallish documents and spreadsheets.) So I guess it would be quite overkill to require 512 MB of RAM for PHP processes for any instance.

**Comment by nautik1** [comment]
I agree it is quite sad to require more memory when it's only for the install process.

**Comment by tlouwers** [comment]
I ran into this memory issue with my hosted instance: here 128 MB is the limit of what I have and I cannot change this.

After upgrading from v17.0.3 to v18.0.3 I tried to install the Community Document Server, which fails with timeouts. Then I switched to using OCC Web, which should not have a timeout. After a short while I get this error (which I guess is exactly as described above):

```
occ $ app:install documentserver_community
The current PHP memory limit is below the recommended value of 512MB.Error: Could not download app documentserver_community
```

**Comment by szaimen** [comment]
Hi, please update to 24.0.8 or better 25.0.2 and report back if it fixes the issue. Thank you!
